### ISSUE TITLE
publish to azure via "mvn azure-functions:deploy" doesn't support service principal

### ISSUE THREAD

**Issue by danyongliu** [microsoft/azure-maven-plugins#125]
Repeat the problem:

1. login to azure through service principal (https://docs.microsoft.com/en-us/cli/azure/authenticate-azure-cli?view=azure-cli-latest#logging-in-with-a-service-principal)
2. Run "mvn azure-functions:deploy" to deploy the function
3. Error like "Failed to execute goal com.microsoft.azure:azure-functions-maven-plugin:0.2.0:deploy (default-cli) on project xxx: java.io.IOException: clientId is null or empty" will be fired

**Comment by jdneo** [comment]
Hi @danyongliu , currently Azure CLi login through service principal is not supported yet. For a work around, you can set the service principal in the Maven's setting.xml.

Check this for more details: https://github.com/Microsoft/azure-maven-plugins/blob/master/docs/common-configuration.md#authentication-method-1-use-the-maven-settingsxml-file

**Comment by jdneo** [comment]
Related issue: https://github.com/Azure/azure-libraries-for-java/issues/417

**Comment by sohelsyed** [comment]
I still see an issue even with settings.xml

What did you pass in plugin configuration ? I am passing below

<plugin>
                    <groupId>com.microsoft.azure</groupId>
                    <artifactId>azure-webapp-maven-plugin</artifactId>
                    <configuration>
                        <authentication>
                            <serverId>azure-auth</serverId>
                        </authentication>
                        <resourceGroup>${functionAppName}</resourceGroup>
                    </configuration>
                </plugin>

**Comment by javisst** [comment]
Hi! Is this still being worked on? We need this for our build pipelines.

**Comment by Flanker32** [comment]
Sorry for the delay
@sohelsyed In plugin configuration, we need to pass the authentication profile id and save the client/tenant/key into maven settings.xml, here is an example.
* settings.xml
```xml
  <servers>
    <server>
      <id>verify</id>
      <configuration>
        <client>${CLIENT ID}</client>
        <tenant>${TENANT ID}</tenant>
        <key>${KEY}</key>
        <environment>AZURE</environment>
      </configuration>
    </server>
  </servers>
```
* maven plugin configuration
``` xml
<plugin>
        <groupId>com.microsoft.azure</groupId>
        <artifactId>azure-webapp-maven-plugin</artifactId>
        <version>1.6.0</version>
        <configuration>
          <authentication>
            <serverId>verify</serverId>
          </authentication>
          ...
```
@javisst Yes, we are still looking for solutions for this issue, before this, please set the service principal in maven's setting file as a workaround.

**Comment by EricZhangzf** [comment]
Hi @Flanker32 ,
I have verified this issue today, it still reproduces on Mac OS, here is my repro steps.
`1. Run: az login --service-principal -u <app-url> -p <password-or-cert> --tenant <tenant> on bash.`
![image](https://user-images.githubusercontent.com/44559777/59909699-e0190800-9442-11e9-9c9b-69eecd6f93ac.png)

`  2.Run: mvn azure-webapp:deploy.`
![image](https://user-images.githubusercontent.com/44559777/59909709-e6a77f80-9442-11e9-843c-0e237565383c.png)

My azure-cli version is 2.0.67, it works well on Windows.

**Comment by EricZhangzf** [comment]
The reason is because some of the caches in the log have not been cleared, and re-generation after clear will solve the problem.

**Comment by Flanker32** [comment]
This was caused by an issue of autorest-clientruntime-for-java and has been fixed in [webapp plugin 1.7.0](https://github.com/microsoft/azure-maven-plugins/releases/tag/azure-webapp-maven-plugin-1.7.0) and [function plugin 1.3.3](https://github.com/microsoft/azure-maven-plugins/releases/tag/azure-functions-maven-plugin-1.3.3)
